File: src/editor/development/app/scripts/event.js
'use strict';

var $event = (function() {
  function arrayHasElement(myArray, searchTerm) {
    for(var i = 0, len = myArray.length; i < len; i++) {
      if (myArray[i] === searchTerm) return i;
    }
    return -1;
  }

  var events = {};
  function EventSystem() { }
  EventSystem.prototype = {
    listen: function(eventName, callback) {
      if(!events.hasOwnProperty(eventName))
        events[eventName] = [];

      events[eventName].push(callback);
    },
    broadcast: function(eventName, eventData) {
      if(events.hasOwnProperty(eventName)) {
        for(var i=0; i < events[eventName].length; ++i) {
          events[eventName][i](eventData);
        }
      } else {
        console.log('warning: broadcast event ' + eventName + ' reached no listeners');
      }
    }
  };
  return new EventSystem();
})();

File: src/server/Event.py
events={}
def listen(eventName, callback):
    if not events.has_key(eventName):
        events[eventName] = []
        pass

    events[eventName].append(callback)
    pass

def broadcast(eventName, eventData):
    if events.has_key(eventName):
        for listener in events[eventName]:
            listener(eventData)
        pass
    else:
        # TODO properly log this problem
        print '[warning] Event "' + eventName + '" reached no listeners!'
    pass

if __name__ == '__main__':
    def tstListener(data):
        from pprint import pprint
        pprint(data)
        pass
    broadcast('test', dict(afterMessage='This should trigger a warning'))
    listen('before', tstListener)
    broadcast('before', dict(beforeMessage='Congrats, you received the before after listening!'))
    listen('after', tstListener)
    broadcast('after', dict(afterMessage='Congrats, you received right on time!'))
    broadcast('after', dict(afterMessage='Congrats, you received right on time!!!'))
    broadcast('before', dict(beforeMessage='Congrats, you received the before after listening (2)!'))
    pass

